# Домашнее задание № 1

### Автор: Кузнецов С.И.

In [1]:
import tensorflow as tf
import tensorflow.keras as keras

Загружаем датасет с https://www.kaggle.com/ajayrana/hymenoptera-data

Датасет предварительно был загружен на Google-диск

In [2]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [3]:
train_generator = keras.preprocessing.image.ImageDataGenerator(rescale=1./225.)
test_generator = keras.preprocessing.image.ImageDataGenerator(rescale=1./225.)

In [4]:
train = train_generator.flow_from_directory('/content/gdrive/My Drive/Colab Notebooks/hymenoptera_data/train',
                                            target_size=(224, 224), 
                                            batch_size=32,
                                            class_mode='binary')
test = train_generator.flow_from_directory('/content/gdrive/My Drive/Colab Notebooks/hymenoptera_data/val',
                                            target_size=(224, 224), 
                                            batch_size=32,
                                            class_mode='binary')

Found 244 images belonging to 2 classes.
Found 153 images belonging to 2 classes.


Создаем сверточную нейронную сеть со сверточными, пуллинговыми и полносвязанными слоями.

In [5]:
i = keras.layers.Input(shape=(224,224,3))
o = keras.layers.Conv2D(128, (5,5), strides=(3,3))(i)
o = keras.layers.MaxPool2D((2,2))(o)
o = keras.layers.ReLU()(o)
o = keras.layers.Conv2D(64, (3,3))(o)
o = keras.layers.MaxPool2D((2,2))(o)
o = keras.layers.ReLU()(o)
o = keras.layers.Conv2D(128, (3,3), strides=(1,1))(o)
o = keras.layers.MaxPool2D((2,2))(o)
o = keras.layers.ReLU()(o)
o = keras.layers.Flatten()(o)
o = keras.layers.Dense(1, activation='sigmoid')(o)

In [6]:
model = keras.Model(inputs=i, outputs=o)
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model.summary()

Model: "functional_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         [(None, 224, 224, 3)]     0         
_________________________________________________________________
conv2d (Conv2D)              (None, 74, 74, 128)       9728      
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 37, 37, 128)       0         
_________________________________________________________________
re_lu (ReLU)                 (None, 37, 37, 128)       0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 35, 35, 64)        73792     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 17, 17, 64)        0         
_________________________________________________________________
re_lu_1 (ReLU)               (None, 17, 17, 64)       

In [7]:
model.fit(train, epochs=10, validation_data=test)

Epoch 1/10
8/8 [==============================] - 3s 320ms/step - loss: 0.7011 - accuracy: 0.4836 - val_loss: 0.6840 - val_accuracy: 0.5425
Epoch 2/10
8/8 [==============================] - 2s 284ms/step - loss: 0.6773 - accuracy: 0.5984 - val_loss: 0.6803 - val_accuracy: 0.5229
Epoch 3/10
8/8 [==============================] - 2s 303ms/step - loss: 0.6470 - accuracy: 0.6434 - val_loss: 0.7316 - val_accuracy: 0.5948
Epoch 4/10
8/8 [==============================] - 2s 288ms/step - loss: 0.6124 - accuracy: 0.6516 - val_loss: 0.6954 - val_accuracy: 0.5556
Epoch 5/10
8/8 [==============================] - 2s 294ms/step - loss: 0.5976 - accuracy: 0.6680 - val_loss: 0.6538 - val_accuracy: 0.6471
Epoch 6/10
8/8 [==============================] - 2s 288ms/step - loss: 0.5633 - accuracy: 0.6967 - val_loss: 0.6660 - val_accuracy: 0.5948
Epoch 7/10
8/8 [==============================] - 2s 297ms/step - loss: 0.5827 - accuracy: 0.6762 - val_loss: 0.6763 - val_accuracy: 0.6013
Epoch 8/10
8/8 [====

В итоге получили точность 0.6144

Добавим слои BatchNormalization к нейросети

In [8]:
i = keras.layers.Input(shape=(224,224,3))
o = keras.layers.BatchNormalization()(i)
o = keras.layers.Conv2D(128, (5,5), strides=(3,3))(o)
o = keras.layers.MaxPool2D((2,2))(o)
o = keras.layers.ReLU()(o)
o = keras.layers.BatchNormalization()(o)
o = keras.layers.Conv2D(64, (3,3))(o)
o = keras.layers.MaxPool2D((2,2))(o)
o = keras.layers.ReLU()(o)
o = keras.layers.BatchNormalization()(o)
o = keras.layers.Conv2D(128, (3,3), strides=(1,1))(o)
o = keras.layers.MaxPool2D((2,2))(o)
o = keras.layers.ReLU()(o)
o = keras.layers.BatchNormalization()(o)
o = keras.layers.Flatten()(o)
o = keras.layers.Dense(1, activation='sigmoid')(o)

In [9]:
model = keras.Model(inputs=i, outputs=o)
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model.summary()

Model: "functional_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_2 (InputLayer)         [(None, 224, 224, 3)]     0         
_________________________________________________________________
batch_normalization (BatchNo (None, 224, 224, 3)       12        
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 74, 74, 128)       9728      
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 37, 37, 128)       0         
_________________________________________________________________
re_lu_3 (ReLU)               (None, 37, 37, 128)       0         
_________________________________________________________________
batch_normalization_1 (Batch (None, 37, 37, 128)       512       
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 35, 35, 64)       

In [10]:
model.fit(train, epochs=10, validation_data=test)

Epoch 1/10
8/8 [==============================] - 3s 322ms/step - loss: 0.8781 - accuracy: 0.5492 - val_loss: 0.6857 - val_accuracy: 0.5948
Epoch 2/10
8/8 [==============================] - 2s 290ms/step - loss: 0.3318 - accuracy: 0.8607 - val_loss: 0.6907 - val_accuracy: 0.5229
Epoch 3/10
8/8 [==============================] - 2s 285ms/step - loss: 0.2224 - accuracy: 0.9180 - val_loss: 0.7139 - val_accuracy: 0.4510
Epoch 4/10
8/8 [==============================] - 2s 295ms/step - loss: 0.1239 - accuracy: 0.9672 - val_loss: 0.7348 - val_accuracy: 0.4641
Epoch 5/10
8/8 [==============================] - 2s 301ms/step - loss: 0.0925 - accuracy: 0.9713 - val_loss: 0.9401 - val_accuracy: 0.4575
Epoch 6/10
8/8 [==============================] - 2s 291ms/step - loss: 0.0440 - accuracy: 1.0000 - val_loss: 0.9042 - val_accuracy: 0.4575
Epoch 7/10
8/8 [==============================] - 2s 290ms/step - loss: 0.0254 - accuracy: 1.0000 - val_loss: 1.0052 - val_accuracy: 0.4575
Epoch 8/10
8/8 [====

Получили точность 0.4575, что ниже, чем в случае без слоев BatchNormalization

Создадим модель ResNet50 и инициализируем ее случайными весами

In [11]:
ResNet50 = keras.applications.ResNet50(include_top=False, weights=None)

In [12]:
i = keras.layers.Input((224, 224, 3), dtype = tf.uint8)
o = tf.cast(i, tf.float32)
o = keras.applications.resnet50.preprocess_input(o)
o = ResNet50(o)
o = keras.layers.Flatten()(o)
o = keras.layers.Dense(1, activation='sigmoid')(o)

In [13]:
model = keras.Model(inputs=i, outputs=o)
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model.summary()

Model: "functional_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_4 (InputLayer)         [(None, 224, 224, 3)]     0         
_________________________________________________________________
tf_op_layer_Cast (TensorFlow [(None, 224, 224, 3)]     0         
_________________________________________________________________
tf_op_layer_strided_slice (T [(None, 224, 224, 3)]     0         
_________________________________________________________________
tf_op_layer_BiasAdd (TensorF [(None, 224, 224, 3)]     0         
_________________________________________________________________
resnet50 (Functional)        (None, None, None, 2048)  23587712  
_________________________________________________________________
flatten_2 (Flatten)          (None, 100352)            0         
_________________________________________________________________
dense_2 (Dense)              (None, 1)                

In [14]:
model.fit(train, epochs=10, validation_data=test)

Epoch 1/10
8/8 [==============================] - 6s 811ms/step - loss: 24.7828 - accuracy: 0.5164 - val_loss: 30468.9941 - val_accuracy: 0.5425
Epoch 2/10
8/8 [==============================] - 5s 567ms/step - loss: 4.4961 - accuracy: 0.5369 - val_loss: 147582.6875 - val_accuracy: 0.5425
Epoch 3/10
8/8 [==============================] - 5s 564ms/step - loss: 2.7872 - accuracy: 0.4877 - val_loss: 61255.0859 - val_accuracy: 0.5425
Epoch 4/10
8/8 [==============================] - 5s 563ms/step - loss: 1.9862 - accuracy: 0.4877 - val_loss: 20622.4805 - val_accuracy: 0.5425
Epoch 5/10
8/8 [==============================] - 5s 568ms/step - loss: 1.6382 - accuracy: 0.4467 - val_loss: 3679.2246 - val_accuracy: 0.5425
Epoch 6/10
8/8 [==============================] - 5s 569ms/step - loss: 3.0936 - accuracy: 0.5451 - val_loss: 3105.9624 - val_accuracy: 0.5425
Epoch 7/10
8/8 [==============================] - 5s 571ms/step - loss: 3.2632 - accuracy: 0.5205 - val_loss: 413.2982 - val_accuracy: 0

Получили точность 0.5425, что хуже, чем в случае с первой нейросетью.

Реализуем предобученную сеть ResNet50 с весами ImageNet.

Добавим 2 полносвязных слоя для повышения качества модели.

In [23]:
ResNet50 = keras.applications.ResNet50(include_top=False, weights='imagenet')

In [28]:
i = keras.layers.Input((224, 224, 3), dtype = tf.uint8)
o = tf.cast(i, tf.float32)
o = keras.applications.resnet50.preprocess_input(o)
o = ResNet50(o)
o = keras.layers.Dense(32, activation='sigmoid')(o)
o = keras.layers.Dense(4, activation='sigmoid')(o)
o = keras.layers.Flatten()(o)
o = keras.layers.Dense(1, activation='sigmoid')(o)

In [29]:
ResNet50.trainable = False

In [30]:
model = keras.Model(inputs=i, outputs=o)
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model.summary()

Model: "functional_11"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_11 (InputLayer)        [(None, 224, 224, 3)]     0         
_________________________________________________________________
tf_op_layer_Cast_5 (TensorFl [(None, 224, 224, 3)]     0         
_________________________________________________________________
tf_op_layer_strided_slice_5  [(None, 224, 224, 3)]     0         
_________________________________________________________________
tf_op_layer_BiasAdd_5 (Tenso [(None, 224, 224, 3)]     0         
_________________________________________________________________
resnet50 (Functional)        (None, None, None, 2048)  23587712  
_________________________________________________________________
dense_11 (Dense)             (None, 7, 7, 32)          65568     
_________________________________________________________________
dense_12 (Dense)             (None, 7, 7, 4)         

In [31]:
model.fit(train, epochs=10, validation_data=test)

Epoch 1/10
8/8 [==============================] - 4s 471ms/step - loss: 0.6958 - accuracy: 0.4918 - val_loss: 0.7036 - val_accuracy: 0.4575
Epoch 2/10
8/8 [==============================] - 3s 355ms/step - loss: 0.6965 - accuracy: 0.4672 - val_loss: 0.6879 - val_accuracy: 0.6405
Epoch 3/10
8/8 [==============================] - 3s 347ms/step - loss: 0.6921 - accuracy: 0.5082 - val_loss: 0.6919 - val_accuracy: 0.4771
Epoch 4/10
8/8 [==============================] - 3s 352ms/step - loss: 0.6824 - accuracy: 0.6148 - val_loss: 0.6823 - val_accuracy: 0.6209
Epoch 5/10
8/8 [==============================] - 3s 351ms/step - loss: 0.6838 - accuracy: 0.5451 - val_loss: 0.6839 - val_accuracy: 0.5686
Epoch 6/10
8/8 [==============================] - 3s 361ms/step - loss: 0.6786 - accuracy: 0.6148 - val_loss: 0.6844 - val_accuracy: 0.5425
Epoch 7/10
8/8 [==============================] - 3s 348ms/step - loss: 0.6750 - accuracy: 0.6311 - val_loss: 0.6799 - val_accuracy: 0.6405
Epoch 8/10
8/8 [====

Получили точность 0.6536

**Выводы:**

1. Максимальная точность модели достигается при использовании предобученной нейросети ResNet50 с добавлением двух полносвязных слоев. Параметры слоев подбирались экспериментально.
2. Добавление слоев с нормализацией (BatchNormalization) ведет к ухудшению качества модели из-за того, что модель переобучается.

